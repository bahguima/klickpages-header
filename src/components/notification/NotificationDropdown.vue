<template>
  <div
    class='nav-notifications'
    :class="{open: notificationsOpen}"
    style="display:block"
    @click="toggleNotifications()"
  >
    <div class="new-notifications-signal"></div>
    <a href="javascript:;" data-toggle="dropdown" class="dropdown-toggle">
      <notification-icon />
    </a>
    <notification-list />
  </div>
</template>

<script>
import NotificationIcon from '../icons/Notification.vue';
import NotificationList from './NotificationList.vue';

export default {
  data() {
    return {
      notificationsOpen: false,
    };
  },
  components: {
    NotificationIcon,
    NotificationList,
  },
  methods: {
    toggleNotifications() {
      this.notificationsOpen = !this.notificationsOpen;
    },
    closeNotifications(event) {
      if (!this.$el.contains(event.target)) {
        this.notificationsOpen = false;
      }
    },
  },
  created() {
    window.addEventListener('click', this.closeNotifications);
  },
  beforeDestroy() {
    window.removeEventListener('click', this.closeNotifications);
  },
};
</script>
