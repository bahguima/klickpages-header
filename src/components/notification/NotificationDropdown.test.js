import { shallowMount, createLocalVue } from '@vue/test-utils';
import NotificationDropdown from './NotificationDropdown.vue';

jest.mock('@/i18n', () => ({
  t: jest.fn((str) => str),
}));

describe('components/notification/NotificationDropdown', () => {
  const localVue = createLocalVue();
  let wrapper;

  describe('when component is mounted', () => {
    describe('and notificationsOpen is truthy', () => {
      beforeAll(() => {
        wrapper = shallowMount(NotificationDropdown, { localVue });
        wrapper.setData({ notificationsOpen: true });
      });

      afterAll(() => {
        wrapper.destroy();
      });

      it('should renders html correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('and notificationsOpen is false', () => {
      beforeAll(() => {
        wrapper = shallowMount(NotificationDropdown, { localVue });
        wrapper.setData({ notificationsOpen: false });
      });

      afterAll(() => {
        wrapper.destroy();
      });

      it('should renders html correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('methods', () => {
    let initialValue;

    describe('toggleNotifications', () => {
      beforeAll(() => {
        wrapper = shallowMount(NotificationDropdown, { localVue });
        initialValue = wrapper.vm.$data.notificationsOpen;
        wrapper.vm.toggleNotifications();
      });

      afterAll(() => {
        wrapper.destroy();
      });
      it('should invert the previous notificationsOpen value', () => {
        expect(wrapper.vm.$data.notificationsOpen).toEqual(!initialValue);
      });
    });

    describe('closeNotifications', () => {
      describe('when the component contains the event target', () => {
        const event = {};

        beforeAll(() => {
          wrapper = shallowMount(NotificationDropdown, { localVue });
          wrapper.setData({ notificationsOpen: true });
          event.target = wrapper.vm.$el;
          wrapper.vm.closeNotifications(event);
        });

        afterAll(() => {
          wrapper.destroy();
        });

        it('should not set to false the notifciationsOpen value', () => {
          expect(wrapper.vm.$data.notificationsOpen).toBeTruthy();
        });
      });

      describe('when the components does not contains the event target', () => {
        const event = {};

        beforeAll(() => {
          wrapper = shallowMount(NotificationDropdown, { localVue });
          wrapper.setData({ notificationsOpen: true });
          event.target = document.body;
          wrapper.vm.closeNotifications(event);
        });

        afterAll(() => {
          wrapper.destroy();
        });

        it('should not set to false the notifciationsOpen value', () => {
          expect(wrapper.vm.$data.notificationsOpen).toBeFalsy();
        });
      });
    });
  });

  describe('lifecycle hooks', () => {
    describe('created()', () => {
      beforeAll(() => {
        window.addEventListener = jest.fn();
        wrapper = shallowMount(NotificationDropdown, { localVue });
      });

      afterAll(() => {
        wrapper.destroy();
      });

      it('should add closeNotifications to the window click event listener', () => {
        expect(window.addEventListener).toHaveBeenCalledWith('click', wrapper.vm.closeNotifications);
      });
    });

    describe('beforeDestroy()', () => {
      beforeAll(() => {
        window.removeEventListener = jest.fn();
        wrapper = shallowMount(NotificationDropdown, { localVue });
        wrapper.destroy();
      });

      afterAll(() => {
        wrapper.destroy();
      });

      it('should remove closeNotifications to the window click event listener', () => {
        expect(window.removeEventListener).toHaveBeenCalledWith('click', wrapper.vm.closeNotifications);
      });
    });
  });
});
