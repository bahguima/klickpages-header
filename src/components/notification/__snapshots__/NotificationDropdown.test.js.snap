// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/notification/NotificationDropdown when component is mounted and notificationsOpen is false should renders html correctly 1`] = `
<div
  class="nav-notifications"
  style="display: block;"
>
  <div
    class="new-notifications-signal"
  />
   
  <a
    class="dropdown-toggle"
    data-toggle="dropdown"
    href="javascript:;"
  >
    <notification-icon-stub />
  </a>
   
  <notification-list-stub />
</div>
`;

exports[`components/notification/NotificationDropdown when component is mounted and notificationsOpen is truthy should renders html correctly 1`] = `
<div
  class="nav-notifications open"
  style="display: block;"
>
  <div
    class="new-notifications-signal"
  />
   
  <a
    class="dropdown-toggle"
    data-toggle="dropdown"
    href="javascript:;"
  >
    <notification-icon-stub />
  </a>
   
  <notification-list-stub />
</div>
`;
